### PR TITLE
Config dependencies

### PR DESCRIPTION
**Pull request by gsquared94**
**Description**
This PR implements [multi config support](http://tinyurl.com/skaffold-multi-configs). Specifically:

1. It allows defining multiple `skaffold.yaml` configurations.
2. Adding dependencies between these configs.
3. Executing named configs via the `-cfg` flag, like:
```
skaffold dev --cfg leeroy-web
```

**User facing changes (remove if N/A)**

Refer to new `integration/examples/multi-config-microservices` example [`README.md`](https://github.com/gsquared94/skaffold/tree/config-dependencies/integration/examples/multi-config-microservices). 

The `skaffold.yaml` now supports a `requires` section that can point to configs in other `skaffold.yaml` files.
```yaml
apiVersion: skaffold/v2beta11
kind: Config
requires:
- path: ./path/to/other/config               #Path to `skaffold.yaml` that needs to be imported
   configs: ["cfg1", "cfg2"]                 #Config to import from above file. If not defined import all configs
   activeProfiles:                                     
   - name: gcb                               #Which profile to activate in the imported configs
     activatedBy: ["gcb"]                    #Any profiles in current config which when activated will activate this profile in the imported config
```
**Follow-up Work (remove if N/A)**

1. Implement remote config support
2. Update documentation

### CODE REVIEW

File: pkg/skaffold/schema/profiles.go [hunk 1]
```diff
@@ -34,28 +34,47 @@ import (
...
+		fallthrough
+	case cfg.RunModes.Render:
+		fallthrough
+	case cfg.RunModes.Run:
+		fallthrough
+	case cfg.RunModes.Diagnose:
+		fallthrough
+	case cfg.RunModes.Delete:
+		c.Profiles = nil
+	}
```
Comment (gsquared94) [inline pkg/skaffold/schema/profiles.go:75]: the current code handles profiles in quite a weird way. In [`profiles.go`](https://github.com/GoogleContainerTools/skaffold/blob/37ed1583747b986c68219a5c8311e72abb4fdcb6/pkg/skaffold/schema/profiles.go#L203) only if at least one profile gets applied then it removes all profile definition objects from the config struct (`cfg.Profiles = nil`). Otherwise it leaves the profile definitions in place. This works out because the `--profile` flag is defined only in a specific selection of run modes (`dev`, `debug`, etc.). So essentially the idea is to retain the profile info if the run mode cares about the config to serialize it as an output (like `generate-pipeline` or `init`).

This switch block makes this expectation explicit.

File: integration/examples/multi-config-microservices/README.md [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+Now, in a different terminal, hit the `leeroy-web` endpoint
+
+```bash
+$ curl $(minikube service leeroy-web --url)
+leeroooooy app!
+```
+
+These are two independently managed instances of Skaffold running on the `leeroy-app` and `leeroy-web` applications. Hitting `Ctrl + C` on each should kill the process and clean up the deployments.
+
+Inorder to iterate on both apps together we reference them as **required** configs in the `multi-config-microservices/skaffold.yaml` file.
```
Comment (MarlonGamez) [inline integration/examples/multi-config-microservices/README.md:40]:
```suggestion
In order to iterate on both apps together we reference them as **required** configs in the `multi-config-microservices/skaffold.yaml` file.
```

**Review by MarlonGamez — COMMENTED** [review]
Took a first pass at this, don't really see any glaring issues. I think it's a good idea to get another pair of eyes on it since there's a lot here. I just have a small typo nit I found

File: pkg/skaffold/schema/validation/validation.go [hunk 3]
```diff
@@ -457,3 +458,16 @@ func validateLogPrefix(lc latest.LogsConfig) []error {
 
 	return nil
 }
+
+func validateSingleKubeContext(configs []*latest.SkaffoldConfig) []error {
```
Comment (nkubala) [inline pkg/skaffold/schema/validation/validation.go:462]: hmm. is it possible to support per-config kube context? i'm not saying we should do it now (that code is a mess), but definitely something to think about. i could imagine scenarios where different components are deployed to different clusters, though maybe this is out of scope for the skaffold workflow.

**Review by nkubala — COMMENTED** [review]
in general the logic looks here correct to my eyes, and the tests certainly look great. left some comments mostly around readability and UX

File: cmd/skaffold/app/cmd/flags.go [hunk 4]
```diff
@@ -67,6 +67,14 @@ var flagRegistry = []Flag{
 		FlagAddMethod: "StringVar",
 		DefinedOn:     []string{"all"},
 	},
+	{
+		Name:          "cfg",
```
[the remark this answers is not in the document]
Comment (gsquared94) [reply cmd/skaffold/app/cmd/flags.go:71]: we already use `--config` for the global config ☹️

File: pkg/skaffold/schema/validation/validation.go (hunk 3 above)
Comment (gsquared94) [reply pkg/skaffold/schema/validation/validation.go:462]: yes, I have left a TODO for that if we ever get to it. We can remove this validation then.

File: cmd/skaffold/app/cmd/parse_config.go [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+	if len(cfgs) == 0 {
+		if len(opts.ConfigurationFilter) > 0 {
+			return nil, fmt.Errorf("did not find any configs matching selection %v", opts.ConfigurationFilter)
+		}
+		return nil, fmt.Errorf("failed to get any valid configs from %s", opts.ConfigurationFile)
+	}
+	return cfgs, nil
+}
+
+func getConfigs(configFile string, configSelection []string, profileSelection []string, opts config.SkaffoldOptions, appliedProfiles map[string]string, requiredConfigs bool, isDependencyConfig bool, configNameToFile map[string]string) ([]*latest.SkaffoldConfig, error) {
```
[the remark this answers is not in the document]
Comment (gsquared94) [reply cmd/skaffold/app/cmd/parse_config.go]: done.

File: pkg/skaffold/schema/profiles.go (hunk 1 above)
Comment (briandealwis) [reply pkg/skaffold/schema/profiles.go:75]: Could this be condensed to a single `case cfg.XX, cfg.YY, cfg.ZZ, ...:`?

**Review by briandealwis — COMMENTED** [review]
Took a very quick run though

File: cmd/skaffold/app/cmd/flags.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (gsquared94) [reply cmd/skaffold/app/cmd/flags.go:71]: I'll rename it to modules

File: pkg/skaffold/schema/profiles.go (hunk 1 above)
Comment (gsquared94) [reply pkg/skaffold/schema/profiles.go:75]: didn't know `case list` existed. Neat

**Review by nkubala — APPROVED** [review]
this LGTM enough to go in now. I'd really love to spend some time thinking about our flag naming, `--config` is now going to be pretty confusing as to what it should do - IMO it should apply here, but instead it will be passing a global config which seems useless.

we don't need to address this here, but definitely something to start thinking about.

**Comment by nkubala** [comment]
cc https://github.com/GoogleContainerTools/skaffold/issues/3150
